lib/Frontend/PrintingDiagnosticConsumer.cpp

```diff
@@ -61,24 +64,572 @@ namespace {
+      out.changeColor(ColoredStream::Colors::CYAN, true);
+      out << "note:";
+      break;
+    case DiagnosticKind::Remark:
+      out.changeColor(ColoredStream::Colors::CYAN, true);
+      out << "remark:";
+      break;
+    }
+    out.resetColor();
+  }
+
+  static void printNumberedGutter(unsigned LineNumber,
+                                  unsigned LineNumberIndent, raw_ostream &Out) {
+    Out.changeColor(ColoredStream::Colors::BLUE, true);
+    Out << llvm::formatv(
+        "{0} | ",
+        llvm::fmt_align(LineNumber, llvm::AlignStyle::Right, LineNumberIndent));
+    Out.resetColor();
+  }
+
+  static void printEmptyGutter(unsigned LineNumberIndent, raw_ostream &Out) {
+    Out.changeColor(ColoredStream::Colors::BLUE, true);
+    Out << std::string(LineNumberIndent + 1, ' ') << "| ";
+    Out.resetColor();
+  }
+
+  // Describe a fix-it out-of-line.
+  static void describeFixIt(SourceManager &SM, DiagnosticInfo::FixIt fixIt,
+                            raw_ostream &Out) {
+    if (fixIt.getRange().getByteLength() == 0) {
+      Out << "[insert '" << fixIt.getText() << "']";
+    } else if (fixIt.getText().empty()) {
+      Out << "[remove '" << SM.extractText(fixIt.getRange()) << "']";
+    } else {
+      Out << "[replace '" << SM.extractText(fixIt.getRange()) << "' with '"
+          << fixIt.getText() << "']";
+    }
+  }
+
+  /// Represents a single line of source code annotated with optional messages,
+  /// highlights, and fix-its.
+  class AnnotatedLine {
+    friend class AnnotatedFileExcerpt;
+
+    // A diagnostic message located at a specific byte in the line.
+    struct Message {
+      unsigned Byte;
+      DiagnosticKind Kind;
+      std::string Text;
+    };
+
+    // A half-open byte range which should be highlighted.
+    struct Highlight {
+      unsigned StartByte;
+      unsigned EndByte;
+    };
+
+    // A half-open byte range which should be replaced with the given text.
+    struct FixIt {
+      unsigned StartByte;
+      unsigned EndByte;
+      std::string Text;
+    };
+
+    unsigned LineNumber;
+    std::string LineText;
+    SmallVector<Message, 1> Messages;
+    SmallVector<Highlight, 1> Highlights;
+    SmallVector<FixIt, 1> FixIts;
+
+    // Adjust output color as needed if this byte is part of a fix-it deletion.
+    void applyStyleForLineByte(unsigned Byte, raw_ostream &Out, bool &Deleted) {
+      bool shouldDelete = false;
+
+      for (auto fixIt : FixIts) {
+        if (Byte >= fixIt.StartByte && Byte < fixIt.EndByte)
+          shouldDelete = true;
+      }
+
+      // Only modify deletions when we reach the start or end of
+      // a fix-it. This ensures that so long as the original
+      // SourceLocs pointed to the first byte of a grapheme cluster, we won't
+      // output an ANSI escape sequence in the middle of one.
+      if (shouldDelete != Deleted) {
+        Out.resetColor();
+        if (shouldDelete) {
+          Out.changeColor(ColoredStream::Colors::RED);
```

Comment:
For accessibility reasons, proposed insertions and deletions shouldn't be distinguished (at least solely) by the colors red and green. This is because red-green colorblindness is the most common form of the condition.

Reply:
Thanks for bringing this up. I've started to address this by adding a textual description of the fix-it to the end of all messages with fix-its, instead of just notes (e.g. `error: argument 'a' must precede argument 'b' [remove ', a: 2' and insert 'a: 2, ']`). 

I think there's more that can and should be done here, but I haven't come up with a design I'm happy with yet, so I'll leave this unresolved for now.

Reply:
I've now modified the formatting to put a `+` below inserted characters and a `-` below deleted characters. The message now looks like this:

<img width="1150" alt="Screen Shot 2020-02-27 at 3 33 17 PM" src="https://user-images.githubusercontent.com/1946221/75496646-a657b400-5976-11ea-96c5-6856bc861b17.png">

Combined with the change above, I think this is more accessible than the first attempt and mostly addresses your second comment as well.